packages/visx-drag/src/util/useStateWithCallback.ts

```diff
@@ -0,0 +1,33 @@
+import { useCallback, useLayoutEffect, useRef, useState } from 'react';
+
+type SetStateWithCallback<State> = (
+  nextState: State | ((currState: State) => State),
+  callback?: (currState: State) => void,
+) => void;
+
+/** A hook that exposes a setState(state, callback) API similar to a component class. */
+export default function useStateWithCallback<State>(
+  initialState: State,
+): [State, SetStateWithCallback<State>] {
+  const [state, setState] = useState(initialState);
+  const callbackRef = useRef<null | ((state: State) => void)>(null);
+
+  const setStateCallback = useCallback<SetStateWithCallback<State>>(
+    (nextState, callback) => {
+      callbackRef.current = callback || null;
+      setState(nextState);
+    },
+    [setState],
+  );
+
+  // if we use useEffect, some callback invocations are skipped
+  useLayoutEffect(() => {
```

Comment:
the hook is the general approach I saw for trying to re-create the `setState(state, callback)` component api in hooks, except they all recommend `useEffect`.  However, when I used `useEffect` I noticed in the `/drag-ii` demo that some of the callbacks were skipped (e.g., the `useDragOptions.onDragStart` [the callback] would be skipped sometimes, which resulted in the previous line being connected to the new line you were starting to draw [`onDragStart` creates a new line in `/drag-ii`]).

I don't have a ton of experience using `useLayoutEffect` vs `useEffect`, I just know that `useLayoutEffect` is blocking while `useEffect` isn't as documented in the react api and we get a warning in `next` now that `useLayoutEffect` cannot be used server-side (for this functionality I don't think it matters).  

Would love any thoughts on this!

Reply:
Hm. Not ideal but this could be something we revisit in the future.